h2o-algos/src/main/java/hex/deeplearning/DeepLearningModel.java

```diff
@@ -32,11 +36,411 @@
  * a scoring history, as well as some helpers to indicate the progress
  */
 
-public class DeepLearningModel extends Model<DeepLearningModel,DeepLearningModel.DeepLearningParameters,DeepLearningModel.DeepLearningModelOutput> implements Model.DeepFeatures {
-  @Override public ToEigenVec getToEigenVec() {
+public class DeepLearningModel extends Model<DeepLearningModel, DeepLearningModel.DeepLearningParameters, DeepLearningModel.DeepLearningModelOutput>
+        implements Model.DeepFeatures, Model.Contributions {
+  @Override
+  public ToEigenVec getToEigenVec() {
     return LinearAlgebraUtils.toEigen;
   }
 
+  class DeepSHAPContributionsWithBackground extends ContributionsWithBackgroundFrameTask<DeepSHAPContributionsWithBackground> {
+
+    final Function<Double, Double> _activation;
+    final Function<Double, Double> _activationDiff;
+    final int[] _origIndices;
+    final int _hiddenLayerMultiplier;
+    final boolean _outputSpace;
+
+    public DeepSHAPContributionsWithBackground(Frame fr, Frame backgroundFrame, boolean perReference, int[] origIndices, boolean outputSpace) {
+      super(fr, backgroundFrame, perReference);
+
+      switch (_parms._activation) {
+        case Tanh:
+        case TanhWithDropout:
+          _activation = this::tanhActivation;
+          // differentials are used only in cases when delta_y/delta_x could be numerically unstable (abs(delta_x)<1e-6)
+          _activationDiff = this::tanhActivationDiff;
+          _hiddenLayerMultiplier = 1;
+          break;
+        case Rectifier:
+        case RectifierWithDropout:
+          _activation = this::rectifierActivation;
+          _activationDiff = this::rectifierActivationDiff;
+          _hiddenLayerMultiplier = 1;
+          break;
+//          exp rectifier is not supported
```

Comment:
Could you remove this commented code, please? Maybe just keep comment the exp rectifier is not supported. Thanks.

Reply:
Thank you for pointing me to this piece of code. I'll throw an exception if ExpRectifier or other unsupported activation functions are used. ExpRectifier seems to be at least partially implemented but it doesn't seem to be supported for training. All currently activation functions supported for training are supported in my implementation of DeepSHAP.